### PR TITLE
[Reassociate] Use uint64_t for repeat count

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch relands #91469 and uses `uint64_t` for repeat count to avoid a miscompilation caused by overflow https://github.com/llvm/llvm-project/pull/91469#discussion_r1623925158.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/Reassociate.cpp [hunk 1]
```diff
@@ -625,10 +532,7 @@ static bool LinearizeExprTree(Instruction *I,
       // Node initially thought to be a leaf wasn't.
       continue;
     assert(!isReassociableOp(V, Opcode) && "Shouldn't be a leaf!");
-    APInt Weight = It->second;
-    if (Weight.isMinValue())
-      // Leaf already output or weight reduction eliminated it.
-      continue;
+    uint64_t Weight = It->second;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/Reassociate.cpp:535]: As we don't reduce weight for nilpotent ops now (e.g., xor), the weight cannot be zero.

**Comment by efriedma-quic** [comment]
I don't see any overflow checks here... is there some reason you can't overflow a uint64_t?

**Comment by dtcxzyw** [comment]
> I don't see any overflow checks here... is there some reason you can't overflow a uint64_t?

I think it overflows when there are more than 2^64 instructions in the IR. Can you provide a counterexample?

**Comment by efriedma-quic** [comment]
I don't have a specific example, but there should probably be a comment with the invariants somewhere.   (If an operand can show up multiple times in the tree, you might run into issues?)

File: llvm/lib/Transforms/Scalar/Reassociate.cpp [hunk 2]
```diff
@@ -562,7 +469,8 @@ static bool LinearizeExprTree(Instruction *I,
                "In leaf map but not visited!");
 
         // Update the number of paths to the leaf.
-        IncorporateWeight(It->second, Weight, Opcode);
+        It->second += Weight;
+        assert(It->second >= Weight && "Weight overflows");
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/Reassociate.cpp:473]: See https://github.com/llvm/llvm-project/blob/374f6554c3e409e4b9b5fd0ec90c5272768f5ab9/llvm/lib/Support/APInt.cpp#L1905-L1909

**Review by nikic — APPROVED** [review]
LGTM. I initially thought we could cause overflow with an exponential tree like https://gist.github.com/nikic/e7736ce11aa66b75ac8fa2b0bf112537. But apparently this is not possible, because any value with multiple uses will be considered as a leaf. That means that the weights can only increase linearly and not exponentially.

**Pull request merged**